File: frontend/src/app/services/event.service.ts
import { Injectable } from '@angular/core';
import {ESLint} from "eslint";
import {Globals} from '../global/globals';
import {HttpClient, HttpParams, HttpResponse, HttpHeaders} from "@angular/common/http";
import {Observable} from "rxjs";
import {EventCreateDto, EventDetailDto, EventEditDto, EventListDto, EventSearchDto} from "../dtos/event";
import {formatIsoDate} from "../util/date-helper";

@Injectable({
  providedIn: 'root'
})
export class EventService {

  private eventBaseUri: string = this.globals.backendUri + "/events";

  constructor(private httpClient: HttpClient, private globals: Globals) {}

  /**
   * Search for events
   *
   * @param searchParams the search parameters
   */
  search(searchParams: EventSearchDto): Observable<EventListDto[]> {
    // TODO: name of searchEventDto is never used? if so, remove it both in back- and frontend

    let params = new HttpParams();
    if (searchParams.earliestStartDate) {
      params = params.append('earliestDate', formatIsoDate(searchParams.earliestStartDate));
    }
    if (searchParams.latestEndDate) {
      params = params.append('latestDate', formatIsoDate(searchParams.latestEndDate));
    }
    if (searchParams.earliestStartTime) {
      params = params.append('earliestStartTime', formatIsoDate(searchParams.earliestStartTime))
    }
    if (searchParams.latestEndTime) {
      params = params.append('latestEndTime', formatIsoDate(searchParams.latestEndTime))
    }
    return this.httpClient.get<EventListDto[]>(this.eventBaseUri + "/search", { params });
  }

  /**
   * Get next upcoming events (5 at most)
   */
  getUpcomingEvents(): Observable<EventListDto[]> {
    return this.httpClient.get<EventListDto[]>(this.eventBaseUri + "/upcoming");
  }

  /**
   * Get an event by its hashId
   *
   * @param hashId the hashId of the event
   * @return an Observable for the event
   */
  getByHashId(hashId: string): Observable<EventDetailDto> {
    let params = new HttpParams().set('hashId', hashId);
    return this.httpClient.get<EventDetailDto>(this.eventBaseUri + '/detail', {params: params});
  }

  /**
   * Delete an event by its hashId
   *
   * @param hashId
   * @return an Observable for the HttpResponse
   */
  delete(hashId: string): Observable<HttpResponse<void>> {
    return this.httpClient.delete<void>(this.eventBaseUri, {observe: 'response', body: hashId});
  }

  /**
   * Create a new event
   *
   * @param eventCreateDto the event to create
   * @return an Observable for the created event
   */
  create(eventCreateDto: EventCreateDto): Observable<Event> {
    return this.httpClient.post<Event>(this.eventBaseUri, eventCreateDto);
  }

  /**
   * Update an event
   *
   * @param eventEditDto the event to update
   * @return an Observable for the updated event
   */
  update(eventEditDto: EventEditDto): Observable<EventEditDto> {
    return this.httpClient.put<EventEditDto>(this.eventBaseUri, eventEditDto);
  }

  uploadIcsFile(file: File): Observable<any> {
    const formData: FormData = new FormData();
    formData.append('file', file, file.name);

    return this.httpClient.post(`${this.eventBaseUri}/import-ics`, formData, {
      headers: new HttpHeaders({
        'Accept': 'application/json'
      })
    });
  }
}

File: backend/src/main/java/at/ac/tuwien/sepr/groupphase/backend/service/EventService.java
package at.ac.tuwien.sepr.groupphase.backend.service;

import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.EventCreateDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.EventDetailDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.EventEditDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.EventListDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.EventSearchDto;
import at.ac.tuwien.sepr.groupphase.backend.exception.IllegalArgumentException;
import at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

/**
 * Service for operations on events (e.g. creating a new Event).
 */
public interface EventService {

    /**
     * Search for events fitting the parameters..
     *
     * @param searchParameters the search parameters
     * @return the list of events that match the search parameters
     */
    List<EventListDto> search(EventSearchDto searchParameters);

    /**
     * Get the upcoming events.
     *
     * @return the list of upcoming events
     */
    List<EventListDto> getUpcomingEvents();

    /**
     * Get the details of an event specified by its Hashed id.
     *
     * @param hashId the Hashed id of the event
     * @return the event details
     * @throws NotFoundException if the event with the given hashId does not exist
     */
    EventDetailDto getByHashId(String hashId) throws NotFoundException;

    /**
     * Create a new event.
     *
     * @param eventCreateDto the event to create
     * @return the created event
     */
    EventCreateDto create(EventCreateDto eventCreateDto);

    /**
     * Update an existing event.
     *
     * @param eventEditDto the event to update
     * @return the updated event
     */
    EventEditDto update(EventEditDto eventEditDto);

    /**
     * Delete an event.
     *
     * @param hashId the Hashed id of the event to delete
     * @throws NotFoundException if the event with the given hashId does not exist
     */
    void delete(String hashId) throws NotFoundException;

    /**
     * Import an ICS file.
     *
     * @param file the ICS file
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.IllegalArgumentException if the file could not be imported
     */
    void importIcsFile(MultipartFile file) throws IllegalArgumentException;
}
